drivers/virtio/virtqueue.c

```diff
@@ -0,0 +1,154 @@
+ */
+
+/*
+ * The maximum queue size is 2^15 (see 2.7),
+ * so any 16bit value larger than that can be used as a sentinel in the next field
+ */
+#define VIRTQ_DESC_NEXT_SENTINEL 0xffff
+
+int virtq_create(struct virtq *v, size_t size)
+{
+	__ASSERT(IS_POWER_OF_TWO(size), "size of virtqueue must be a power of 2");
+	__ASSERT(size <= KB(32), "size of virtqueue must be at most 32KB");
+	/*
+	 * For sizes and alignments see table in spec 2.7. We are supporting only modern virtio, so
+	 * we don't have to adhere to additional constraints from spec 2.7.2
+	 */
+	size_t descriptor_table_size = 16 * size;
+	size_t available_ring_size = 2 * size + 6;
+	size_t used_ring_pad = (descriptor_table_size + available_ring_size) % 4;
+	size_t used_ring_size = 8 * size + 6;
+	size_t shared_size =
+		descriptor_table_size + available_ring_size + used_ring_pad + used_ring_size;
+	size_t v_size = shared_size + sizeof(struct virtq_receive_callback_entry) * size;
+
+	uint8_t *v_area = k_aligned_alloc(16, v_size);
+
+	if (!v_area) {
+		LOG_ERR("unable to allocate virtqueue");
+		return -ENOMEM;
+	}
+
+	v->num = size;
+	v->desc = (struct virtq_desc *)v_area;
+	v->avail = (struct virtq_avail *)((uint8_t *)v->desc + descriptor_table_size);
+	v->used = (struct virtq_used *)((uint8_t *)v->avail + available_ring_size + used_ring_pad);
+	v->recv_cbs = (struct virtq_receive_callback_entry *)((uint8_t *)v->used + used_ring_size);
+
+	/*
+	 * At the beginning of the descriptor table, the available ring and the used ring have to be
+	 * set to zero. It's the case for both PCI (4.1.5.1.3) and MMIO (4.2.3.2) transport options.
+	 * Its unspecified for channel I/O (chapter 4.3), but its used on platforms not supported by
+	 * Zephyr, so we don't have to handle it here
+	 */
+	memset(v_area, 0, v_size);
+
+	k_mutex_init(&v->mtx);
+	v->last_used_idx = 0;
+
+	k_stack_alloc_init(&v->free_desc_stack, size);
+	for (uint16_t i = 0; i < size; i++) {
+		k_stack_push(&v->free_desc_stack, i);
+	}
+
+	return 0;
+}
+
+void virtq_free(struct virtq *v)
+{
+	k_free(v->desc);
+	k_stack_cleanup(&v->free_desc_stack);
+}
+
+static int virtq_add_available(struct virtq *v, uint16_t desc_idx)
+{
+	uint16_t new_idx_le = sys_cpu_to_le16(sys_le16_to_cpu(v->avail->idx) % v->num);
+
+	v->avail->ring[new_idx_le] = sys_cpu_to_le16(desc_idx);
+	barrier_dmem_fence_full();
+	v->avail->idx = sys_cpu_to_le16(sys_le16_to_cpu(v->avail->idx) + 1);
+
+	return 0;
+}
+
+int virtq_add_buffer_chain(
+	struct virtq *v, struct virtq_buf *bufs, uint16_t bufs_size,
+	uint16_t device_readable_count, virtq_receive_callback cb, void *cb_opaque)
+{
+	uint64_t total_len = 0;
+
+	for (int i = 0; i < bufs_size; i++) {
+		total_len += bufs[i].len;
+	}
+
+	/* According to the spec 2.7.5.2 the maximum size of descriptor chain is 4GB */
+	if (total_len > GB(4)) {
+		LOG_ERR("buffer chain is longer than 2^32 bytes");
+		return -EINVAL;
+	}
+
+	k_mutex_lock(&v->mtx, K_FOREVER);
```

Comment:
Is _virtq_add_buffer_chain()_ intended to ever be called from an ISR? As it is written, it must not be called from an ISR due to both the use of _K_FOREVER_ as a timeout parameter and the use of _k_mutex_lock()_. If not, then a comment in the routine's description stating that restriction.

Reply:
Thanks for pointing this out. While the only device we implemented so far (virtiofs) doesn't need it, we've analyzed the other virtio devices and some of them may benefit from/require placing descriptors back in the available ring in the ISR directly after receiving them (mainly the devices with dedicated rx queue, where we would like to keep the available ring full and would possibly have no other place to call `virtq_add_buffer_chain`).

With the above in mind we modified this function to take a timeout param and changed mutex to a spinlock, to allow calling it from ISR with `K_NO_WAIT`.